pkg/operator/ceph/config/keyring/cephx.go

```diff
@@ -0,0 +1,183 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keyring
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	v1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/operator/ceph/version"
+)
+
+var CephAuthRotateSupportedVersion = version.CephVersion{Major: 20, Minor: 2, Extra: 0}
+
+// CephxKeyIdentifierAnnotation is the annotation that should be applied to pod specs to
+// ensure that pods restart after keys are rotated (and not restarted when keys are not rotated).
+// The keyring secret resourceVersion is suggested but not always available.
+//
+//nolint:gosec // G101: this is not hardcoded credentials
+const CephxKeyIdentifierAnnotation = "cephx-key-identifier"
+
+// ShouldRotateCephxKeys determines whether CephX keys should be rotated based on the CephX key
+// rotation config, the version of Ceph present in the image being deployed (desiredCephVersion),
+// and the last-reconciled CephX key status.
+// runningCephVersion is used to determine if the cluster is capable of rotating CephX keys.
+// Intended to use running/desired ceph version from CurrentAndDesiredCephVersion().
+func ShouldRotateCephxKeys(cfg v1.CephxConfig, runningCephVersion, desiredCephVersion version.CephVersion, status v1.CephxStatus) (bool, error) {
+	if !runningCephVersion.IsAtLeast(CephAuthRotateSupportedVersion) {
+		logger.Debugf("should not rotate cephx keys using unsupported ceph version %#v", runningCephVersion)
+		return false, nil
+	}
+
+	if status.KeyCephVersion == v1.UninitializedCephxKeyCephVersion {
+		return false, nil // no need to rotate key when key isn't yet initialized
+	}
+
+	switch cfg.KeyRotationPolicy {
+	case v1.CephxKeyRotationPolicy(""), v1.DisabledCephxKeyRotationPolicy:
+		return false, nil
+	case v1.KeyGenerationCephxKeyRotationPolicy:
+		return cfg.KeyGeneration > status.KeyGeneration, nil
+	case v1.WithCephVersionUpdateCephxKeyRotationPolicy:
+		if version.IsIdentical(desiredCephVersion, version.CephVersion{}) {
+			// likely cause of this is developer error. if that makes it to release, it's probably
+			// best to not cause unrecoverable error for users
+			logger.Info("ShouldRotateCephxKeys(): desiredCephVersion is unspecified")
+			return false, nil
+		}
+
+		if status.KeyCephVersion == "" {
+			return true, nil // when previous version is unknown, assume rotation
+		}
+
+		statusVer, err := parseCephVersionFromStatusVersion(status.KeyCephVersion)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to determine if cephx keys need to be rotated under %q rotation policy; failed to parse key ceph version from status %q", cfg.KeyRotationPolicy, status.KeyCephVersion)
+		}
+		// by API spec, commit ID is not part of CephCluster.status.version.version or
+		// keyCephVersion, so strip it from version found in ceph image
+		desiredCephVersion.CommitID = ""
+		if version.IsSuperior(desiredCephVersion, statusVer) {
```

Comment:
Do we need to check for superior version? Perhaps a string compare for a changing image is sufficient, and it might remove some code for version parsing. 

Or we could just return true if there is any change, whether superior or inferior.